### ISSUE TITLE
Proof of DOT

### ISSUE THREAD

**Issue by eskimor** [paritytech/polkadot-sdk#6173]
This ticket is about adding a generic sybil resistance functionality to Polkadot to be used permissionlessly by any protocol that might need it. This includes Polkadot native infrastructure, but given its permissionlessness, it can be used by any (p2p) protocol to enhance its resilience.

The idea is not new and rather simple: We use a proof of stake mechanism. With this we can prefer participants with stake over participants without. Acquiring stake becomes expensive very quickly if you want to attack. This mechanism can be added to any protocol in a backwards compatible way: Existing clients would still be served based on best-effort (same as they did before), but in addition, participants can lock up DOT for better quality of service, especially in attack scenarios.

## Ingredients

- A pallet providing the locking mechanism: pallet-pod + runtime API
- A runtime API to provide a nice interface for querying
- A library/facade to access this system independently of where the functionality actually lives
- RFC for adding this functionality to a system chain. people chain seems ideal.

#### Extrinsics

```rust

type ProtocolName = BoundedVec<u8, MAX_PROTOCOL_NAME_SIZE>;

struct ProtocolConfiguration {

  /// Number of participants (who will be willing to lock up some DOT) do you expect to 
  /// exist for your protocol. Required deposits will be relatively cheap, if the number of participants stays around that number, but will grow to something larger than the entire DOT supply, once number of participants approach ten times that value.
  /// Ballpark estimate is enough (order of magnitude). E.g. 10, 100, 1000, 10_0000, ...
  /// assert!(expected_participants >= 10 && expected_participants <= MAX_PARTICIPANTS);
  expected_participants: u32,
  
  /// The period, participants have to wait to get their funds unlocked, after they decided to   
  /// leave. 
  unlock_period: BlockNumber,
}

fn register_protocol(origin, protocol_name: ProtocolName, protocol: ProtocolConfiguration);

```

Register a new protocol you want to have sybil resiistance for, with a name and an expected number of participants. 

origin can be any, including root or a chain. Some fee will be charged and some DOT locked for the lifetime of the registered protocol. The name must be unique.

```rust
fn participate(origin, protocol_name, max_lock_amount: Balance, peer_identifier: PeerIdentifier);
```

Show intent to become a participant in the protocol, by registering some identifier (e.g. your PeerId). You can specify a maximum amount of DOT you are willing to have locked for this. Actual amount locked will be determined based on current occupation.

```rust
fn leave(origin, protocol_name,  peer_identifier);
```

Leave the protocol: Your peer identifier will be removed and you get your funds unlocked after `unlock_period` blocks.

```rust
fn remove_protocol(origin, protocol: ProtocolName);
```

Remove a protocol and claim back the storage deposit for the protocol. This call will fail if there are still participants in the network. Hence, when registering, you should deal with the possibility that you might never get your storage deposit back. The data stored for a protocol is only the configuration, thus that amount should be small. I believe this is acceptable and provides the benefit that the system is by default decentralized: Anybody can create a protocol, but the resulting system would still be decentralized, as nobody, not even that creator will be able to stop it.

#### View functions

The pallet will also expose the following functions, which should be exported as a runtime API:

```rust
type Participants = BoundedVec<PeerIdentifier, MAX_PARTICIPANTS>;

fn protocol_participants(protocol: ProtocolName) -> Result<Participants>;
```

`MAX_PARTICIPANTS` should be constrained in a way, such that this call is feasible. We could use a cursor based accessor function, but given that this is supposed to be called from a runtime API, I don't think this complexity is necessary (For now. We can add that later once we actually have demand for networks that large).

Note also that:

- Above type is not how we store the data (POV size stays small).
- For networks that grow too much, one can always create another protocol. E.g. "my_fancy_network/0", "my_fancy_network/1". This also means that `expected_participants` can be set conservatively, as new protocols can always be added.
- Returning a few million entries this way should be feasible in a runtime API. 1 PeerId (without prefix) is 32 bytes, thus 10 million entries would be roughly 320MB - which works. Once we get into that ballpark, we should start looking into getting more fancy. E.g. to only retrieve changes from the last state or only storing a root on chain and keeping the actual data off-chain - efficiently query able, ... Having a simple API to start with and also later on for smallish networks seems sensible though.

### Implementation

Protocol registration will ensure that no other protocol with the same name exists. Then it will store the configuration under the protocol's name.

Participation is more interesting. We need to ensure that the system is not itself DOSable. In particular it must not be possible for an attacker to get honest nodes lose their spot, yet we still have limited capacity. We achieve this by setting a conceptual limit on the number of participants to 

```
limit = 10 * expected_participants
```

further we set the following:

```
base_price = 0.1 DOT // (pallet configuration)
step = limit/40 // (pallet configuration)
```

Then we calculate the deposit we take for the `n`th  participant via:

```
deposit(n) = base_price * 2^(n/step) // integer arithmetic
```

Example, at the time of registration there have been already 100 participants registered, assuming `limit = 1000`, then the deposit for the `participate` call would be: `0.1 * 2^(100/(1000/40)) = 1.6 DOT`

Parameters have been picked such that you can not reach `limit`, as the amount of DOT, you would need to lock, exceed the entire DOT supply. (Economically bounded)

Cost of censorship attacks are high and in favor of the defender. For doubling the price, the attacker would need to invest `step/2` times that doubled price. Trying to bring up the price by a factor of 8, will cost x *( step + 2* step + 4* step). Where `x` was the current price of the time of the attack. This results in 175 * x for the attacker to pay, while the to be censored node will need to pay 8*x. (We assume limit to be 1000 and step size thus 25). For smaller limits the attacker would still need to pay at least the same as the to be censored node.

For each participant we store in a

```rust
#[pallet::storage]
pub type Participants<T: Config> = StorageDoubleMap<ProtocolName, PeerId, T::AccountId>;
```
the `Participant` data:

```rust
struct ParticipantAccount {
  owner: Account,
  locked: Balance,
}
```
For `leaving` we will remove the entry in above storage map and note in `PendingRemoval` storage that the funds may be reclaimed at `now + unlocking_period`. We either check in `on_initialize` for accounts that should get their funds back or we add an additional extrinsic for claiming back funds that passed the unlocking period.

## Provided Facade

To make all of this easily accessible we need to provide a Rust library with the following entry points:

```rust
fn register_protocol(origin, protocol_name: ProtocolName, protocol: ProtocolConfiguration);
fn participate(origin, protocol_name, max_lock_amount: Balance, peer_identifier);
fn leave(origin, protocol_name,  peer_identifier: PeerId);
fn remove_protocol(origin, protocol: ProtocolName);
fn protocol_participants(protocol: ProtocolName) -> Result<Participants>;
```

So essentially the functionality provided by the pallet. With a few key differences:

- We can use more concrete types. E.g. use an actual `PeerId`, while in the pallet we will use some BoundedVec<u8> to keep it protocol agnostic. 
- The calls should just work. The library will discover the Polkadot network on its own with minimal configuration and more importantly will craft extrinsics for the right chain for the user to sign and will also know by itself what chain to query to retrieve protocol participants. 

This will allow us to move the functionality around (if necessary), change implementation details and even allows for sharding if that ever became necessary. The Facade would just determine what shard to lookup a given protocol based on the name.

**Comment by skunert** [comment]
Nice proposal!

I expect that its not super very cheap to fetch information from this facade, so for Peer related info and generally operations that happen often, we would need to cache the list of participants locally for sure.

**Comment by burdges** [comment]
> A pallet providing the locking mechanism: pallet-pod + runtime API

It's a dedicated lock, not an overlapping lock ala restaking?  We've typically used overlapping locks ala restaking, but in general this really depends upon what your doing.

> I expect that its not super very cheap to fetch information from this facade, so for Peer related info and generally operations that happen often, we would need to cache the list of participants locally for sure.

In the worse case, remove Merkle proofs would not necessarily be more expensive than the local Merkle proofs required by local state accesses, just depends how heavily used.  In fact, the remove Merkle proofs could even become cheaper than local ones, assuming the majority of state accesses were remote.

**Comment by eskimor** [comment]
It must be without overlap. Without an attack the amounts locked will stay low anyway, but if there is one, the economic pressure must go up to have an effect.

> I expect that its not super very cheap to fetch information from this facade, so for Peer related info and generally operations that happen often, we would need to cache the list of participants locally for sure.

Yes. This is one of the reasons for the `unlock_period`. The idea is you only update once in a while. E.g. once per day/ once per session. Whatever makes sense for the application.

**Comment by eskimor** [comment]
Scaling to billions, proposal by Claude - not everything makes sense, but is a starting point:

# Proof-of-DOT Billion Scale: Detailed Implementation Design

## Core Architecture

### The Two-Layer Design

**Layer 1: On-Chain Accounting (Source of Truth)**
```rust
// Full participant information - CANNOT be replaced by filters
StorageDoubleMap<ProtocolId, PeerId, ParticipantInfo> where:
  ParticipantInfo {
    account: AccountId,        // Who owns this registration
    deposit: Balance,          // How much they locked
    registered_at: BlockNumber, // When they joined
    lock_until: BlockNumber,   // When they can withdraw
  }
```

This is **absolutely required** for:
- Knowing who can withdraw funds
- Calculating unlock times
- Handling slashing/challenges
- Preventing double registration
- Refunding exact amounts

At billion scale: 1B entries × ~100 bytes = **100GB on-chain storage**

With NOMT (Near-zero Overhead Merkle Trees), this state size is manageable for Polkadot parachains.

**Layer 2: Cuckoo Filter Cache (For Provider Efficiency)**
```rust
// Derived from Layer 1, optimized for membership queries
StorageMap<(ProtocolId, ShardId), CuckooFilterData> where:
  CuckooFilterData {
    filter: Vec<u8>,           // ~6MB per shard
    version: BlockNumber,      // When last updated
    item_count: u32,          // For load balancing
  }
```

This enables:
- Fast local membership checks (50ns)
- Efficient updates (only affected shards)
- Bandwidth optimization for providers (1.5GB total vs 100GB)

## API Design (Practical for Billion Scale)

```rust
// Runtime API - CRITICAL: No function returns more than a few MB!
sp_api::decl_runtime_apis! {
  pub trait ProofOfDotApi {
    // ❌ NEVER DO THIS - Would return 40GB+!
    // fn get_all_participants(protocol_id: ProtocolId) -> Vec<(PeerId, AccountId)>;

    // ✅ Paginated access to raw participant data (when absolutely needed)
    fn get_participants_page(
      protocol_id: ProtocolId,
      start_key: Option<PeerId>,
      limit: u32,  // Max 1000
    ) -> ParticipantPage {
      participants: Vec<(PeerId, ParticipantInfo)>,
      next_key: Option<PeerId>,
      total_count: u32,
    }

    // ✅ Get single membership shard (~6MB max)
    fn get_membership_shard(
      protocol_id: ProtocolId,
      shard_id: u8
    ) -> Option<CuckooFilterData>;

    // ✅ Get shard metadata only (256 * 16 bytes = 4KB)
    fn get_shard_metadata(
      protocol_id: ProtocolId
    ) -> Vec<ShardMetadata> {
      shard_id: u8,
      version: BlockNumber,
      item_count: u32,
      size_bytes: u32,
    }

    // ✅ Get recent updates (bounded list)
    fn get_updates_since(
      protocol_id: ProtocolId,
      since_block: BlockNumber,
      max_updates: u32,  // Max 10000
    ) -> UpdateBatch {
      updates: Vec<IdentityUpdate>,
      from_block: BlockNumber,
      to_block: BlockNumber,
      has_more: bool,
    }

    // ✅ Check single membership (for verification)
    fn is_participant(
      protocol_id: ProtocolId,
      peer_id: PeerId
    ) -> bool;

    // ✅ Get pending registrations for a shard (bounded)
    fn get_pending_registrations(
      protocol_id: ProtocolId,
      shard_id: u8,
      limit: u32,  // Max 1000
    ) -> Vec<PeerId>;

    // ✅ Get protocol statistics (small fixed size)
    fn get_protocol_stats(protocol_id: ProtocolId) -> ProtocolStats {
      total_participants: u32,
      total_locked: Balance,
      pending_registrations: u32,
      last_batch_block: BlockNumber,
      shards_active: u8,
    }
  }
}

enum IdentityUpdate {
  Added { peer_id: PeerId, block: BlockNumber },
  Removed { peer_id: PeerId, block: BlockNumber },
}
```

## Provider Synchronization Strategy

### Initial Sync (Efficient Approach)
```rust
async fn initial_sync(&mut self, protocol_id: ProtocolId) {
  // Step 1: Get metadata first (4KB)
  let metadata = runtime_api.get_shard_metadata(protocol_id).await?;

  println!("Protocol has {} participants across {} shards",
           metadata.iter().map(|m| m.item_count).sum::<u32>(),
           metadata.len());

  // Step 2: Determine which shards we need
  let needed_shards = determine_needed_shards(&metadata);

  // Step 3: Download shards in parallel with rate limiting
  let semaphore = Arc::new(Semaphore::new(5)); // Max 5 concurrent

  let futures = needed_shards.iter().map(|shard_id| {
    let sem = semaphore.clone();
    async move {
      let _permit = sem.acquire().await?;
      let data = runtime_api.get_membership_shard(protocol_id, *shard_id).await?;
      Ok::<_, Error>((*shard_id, data))
    }
  });

  // Total download: ~1.5GB for all shards, ~6MB per shard
  let results = join_all(futures).await;

  // Step 4: Store locally
  for result in results {
    let (shard_id, data) = result?;
    let filter = CuckooFilter::decode(&data.filter)?;
    self.shards.insert(shard_id, filter);
    self.versions.insert(shard_id, data.version);
  }
}
```

### Incremental Updates (Bandwidth Efficient)
```rust
async fn sync_updates(&mut self, protocol_id: ProtocolId) {
  let mut last_sync_block = self.last_sync_block;

  loop {
    sleep(Duration::from_secs(30)).await;

    // Step 1: Get metadata to check for changes (4KB)
    let metadata = runtime_api.get_shard_metadata(protocol_id).await?;

    // Step 2: Identify changed shards
    let mut changed_shards = vec![];
    for shard_meta in metadata {
      if let Some(&our_version) = self.versions.get(&shard_meta.shard_id) {
        if shard_meta.version > our_version {
          changed_shards.push(shard_meta.shard_id);
        }
      }
    }

    // Step 3: Fetch only changed shards
    for shard_id in changed_shards {
      let data = runtime_api.get_membership_shard(protocol_id, shard_id).await?;
      let filter = CuckooFilter::decode(&data.filter)?;

      self.shards.insert(shard_id, filter);
      self.versions.insert(shard_id, data.version);

      log::info!("Updated shard {} ({}MB)", shard_id, data.filter.len() / 1_000_000);
    }

    // Step 4: Alternative - get incremental updates
    if self.supports_incremental {
      let updates = runtime_api.get_updates_since(
        protocol_id,
        last_sync_block,
        10000  // Max updates per batch
      ).await?;

      // Apply updates to local filters
      for update in updates.updates {
        match update {
          IdentityUpdate::Added { peer_id, .. } => {
            let shard_id = peer_id.as_bytes()[0];
            if let Some(filter) = self.shards.get_mut(&shard_id) {
              filter.insert(&peer_id);
            }
          }
          IdentityUpdate::Removed { peer_id, .. } => {
            let shard_id = peer_id.as_bytes()[0];
            if let Some(filter) = self.shards.get_mut(&shard_id) {
              filter.delete(&peer_id);
            }
          }
        }
      }

      last_sync_block = updates.to_block;
    }
  }
}
```

### Lazy Loading Strategy (For Light Providers)
```rust
struct LazyProvider {
  // Only download shards as needed
  loaded_shards: HashMap<u8, CuckooFilter>,
  shard_cache: LruCache<u8, CuckooFilter>,  // Keep last 10 shards
}

async fn check_membership_lazy(&mut self, peer_id: &PeerId) -> bool {
  let shard_id = peer_id.as_bytes()[0];

  // Check if we have this shard
  if let Some(filter) = self.loaded_shards.get(&shard_id) {
    return filter.contains(peer_id);
  }

  // Check cache
  if let Some(filter) = self.shard_cache.get(&shard_id) {
    return filter.contains(peer_id);
  }

  // Need to fetch this shard (6MB)
  if let Some(data) = runtime_api.get_membership_shard(protocol_id, shard_id).await? {
    let filter = CuckooFilter::decode(&data.filter)?;
    let result = filter.contains(peer_id);

    // Cache for future use
    self.shard_cache.put(shard_id, filter);

    return result;
  }

  false
}
```

## Complete Registration Flow

### Step 1: User Registration
```rust
fn participate(origin, protocol_id: ProtocolId, peer_id: PeerId) {
  let who = ensure_signed(origin)?;

  // Calculate required deposit
  let participant_count = ParticipantCounts::<protocol_id>::get();
  let deposit = calculate_deposit(participant_count, protocol_params);

  // Lock tokens
  T::Currency::reserve(&who, deposit)?;

  // Store full information (Layer 1)
  Participants::<protocol_id, peer_id>::insert(ParticipantInfo {
    account: who.clone(),
    deposit,
    registered_at: current_block,
    lock_until: current_block + protocol_params.lock_period,
  });

  // Add to pending batch (for Layer 2 update)
  let shard_id = peer_id.as_bytes()[0];
  PendingAdditions::<protocol_id, shard_id>::append(peer_id);

  // Update counts
  ParticipantCounts::<protocol_id>::mutate(|c| *c += 1);

  emit(Event::ParticipantRegistered { protocol_id, peer_id, who, deposit });
}
```

### Step 2: Batch Processing (Every N Blocks)
```rust
fn on_finalize(block: BlockNumber) {
  if block % BATCH_INTERVAL != 0 { return; }

  for protocol_id in RegisteredProtocols::iter() {
    let mut updated_shards = vec![];

    for shard_id in 0..256 {
      let additions = PendingAdditions::<protocol_id, shard_id>::take();
      let removals = PendingRemovals::<protocol_id, shard_id>::take();

      if additions.is_empty() && removals.is_empty() { continue; }

      // Update cuckoo filter for this shard
      let mut filter = CuckooFilters::<protocol_id, shard_id>::get()
        .unwrap_or_else(|| CuckooFilter::with_capacity(4_000_000));

      for peer_id in &additions {
        filter.insert(peer_id)?;
      }

      for peer_id in &removals {
        filter.delete(peer_id)?;
      }

      // Store updated filter
      CuckooFilters::<protocol_id, shard_id>::insert(CuckooFilterData {
        filter: filter.encode(),
        version: block,
        item_count: filter.len(),
      });

      updated_shards.push(shard_id);

      // Store update history for incremental sync
      for peer_id in additions {
        UpdateHistory::<protocol_id>::append(
          IdentityUpdate::Added { peer_id, block }
        );
      }
      for peer_id in removals {
        UpdateHistory::<protocol_id>::append(
          IdentityUpdate::Removed { peer_id, block }
        );
      }
    }

    // Emit event with updated shards
    if !updated_shards.is_empty() {
      emit(Event::ShardsUpdated {
        protocol_id,
        shard_ids: updated_shards,
        block
      });
    }
  }
}
```

## Corner Cases and Solutions

### 1. Provider Wants Full Participant List
**Problem**: Some use cases might need the full list (analytics, audit).

**Solution**: Use pagination, never return everything at once:
```rust
async fn download_all_participants(&self, protocol_id: ProtocolId) -> Vec<(PeerId, ParticipantInfo)> {
  let mut all_participants = vec![];
  let mut start_key = None;

  loop {
    let page = runtime_api.get_participants_page(
      protocol_id,
      start_key,
      1000  // Page size
    ).await?;

    all_participants.extend(page.participants);

    if let Some(next) = page.next_key {
      start_key = Some(next);

      // Rate limit to avoid overwhelming the node
      sleep(Duration::from_millis(100)).await;
    } else {
      break;
    }

    log::info!("Downloaded {}/{} participants",
              all_participants.len(),
              page.total_count);
  }

  all_participants
}
```

### 2. Mass Registration Event
**Problem**: 1 million registrations in a short period.

**Solution**:
- Pending lists have bounds (e.g., max 10K per shard)
- Excess goes into overflow queue
- Process in multiple batches
- Providers see registrations in pending list immediately

### 3. Filter Rebuild Required
**Problem**: Filter corrupted or needs resizing.

**Solution**:
```rust
fn rebuild_filter(protocol_id: ProtocolId, shard_id: u8) {
  // Iterate over primary storage (paginated internally)
  let mut filter = CuckooFilter::with_capacity(5_000_000);

  // Use internal iteration to avoid loading all at once
  Participants::<protocol_id>::iter_prefix_keys(shard_id)
    .for_each(|peer_id| {
      filter.insert(&peer_id);
    });

  CuckooFilters::<protocol_id, shard_id>::insert(CuckooFilterData {
    filter: filter.encode(),
    version: current_block,
    item_count: filter.len(),
  });
}
```

### 4. Provider Network Partitioned During Mass Updates
**Problem**: Provider offline, missed many updates.

**Solution**: Three recovery strategies:
```rust
// Strategy 1: Incremental catch-up (if not too far behind)
if blocks_behind < 10000 {
  let updates = runtime_api.get_updates_since(last_known_block, 10000).await?;
  apply_updates(updates);
}

// Strategy 2: Re-download changed shards (if moderately behind)
else if blocks_behind < 100000 {
  let metadata = runtime_api.get_shard_metadata(protocol_id).await?;
  for shard in metadata {
    if shard.version > our_version[shard.shard_id] {
      download_shard(shard.shard_id).await?;
    }
  }
}

// Strategy 3: Full resync (if very far behind)
else {
  initial_sync().await?;
}
```

### 5. False Positive Verification
**Problem**: Cuckoo filter says registered, but we need certainty.

**Solution**: Single-query verification:
```rust
async fn verify_critical_operation(&self, peer_id: &PeerId) -> bool {
  // Fast path: check local filter
  if !self.is_registered_locally(peer_id) {
    return false;  // Definitely not registered
  }

  // Slow path: verify on-chain for critical operations
  runtime_api.is_participant(protocol_id, peer_id).await
}
```

## Efficiency Analysis

### Data Transfer Sizes
| Operation | Data Size | Frequency |
|-----------|-----------|-----------|
| Get shard metadata | 4KB | Every 30s |
| Get single shard | 6MB | On change |
| Get all shards | 1.5GB | Initial sync only |
| Get updates batch | <1MB | Every 30s |
| Get participant page | 100KB | Rare (audit) |
| Check single membership | 100 bytes | On demand |

### Bandwidth Requirements
- **Initial sync**: 1.5GB one-time
- **Steady state**: ~10MB/hour (metadata + occasional shard updates)
- **During high activity**: ~150MB/hour (multiple shard updates)
- **Never**: 40GB+ transfers

### Query Performance
- **Local membership check**: 50ns
- **Metadata fetch**: ~10ms
- **Single shard fetch**: ~100ms (6MB transfer)
- **Single membership verification**: ~50ms (one RPC call)

## Critical Design Principles

1. **Never return unbounded data** - Every API call has explicit limits
2. **Pagination for large datasets** - Use cursors and page sizes
3. **Metadata before data** - Check versions before downloading
4. **Incremental updates when possible** - Avoid re-downloading unchanged data
5. **Rate limiting built-in** - Prevent overwhelming nodes with requests
6. **Graceful degradation** - Multiple strategies for different scenarios

## Implementation Recommendations

### For the Runtime API
1. Enforce maximum return sizes (e.g., 10MB per call)
2. Always provide pagination for collections
3. Include version/timestamp metadata
4. Support both full and incremental sync patterns

### For Provider Libraries
```rust
pub struct ProviderConfig {
  // Sync strategy
  sync_mode: SyncMode,  // Full, Lazy, Incremental

  // Resource limits
  max_memory: usize,    // Max memory for filters (default 2GB)
  max_bandwidth: usize, // Max bandwidth per hour (default 1GB)

  // Performance tuning
  update_interval: Duration,  // How often to check for updates
  parallel_downloads: usize,  // Max concurrent shard downloads

  // Reliability
  retry_strategy: RetryStrategy,
  fallback_nodes: Vec<String>,
}

enum SyncMode {
  Full,        // Download all shards upfront
  Lazy,        // Download shards on demand
  Incremental, // Start full, then incremental updates
  Regional,    // Only specific shards
}
```

## Conclusion

With proper API design that respects bandwidth limitations, the sharded cuckoo filter approach scales to billions of users while keeping data transfer practical. No function ever returns more than a few MB, and providers can choose their sync strategy based on their resources and requirements. The 100GB on-chain state is manageable with NOMT, and the 1.5GB filter cache enables fast local lookups without overwhelming the network.

**Comment by burdges** [comment]
I missed the "proposal by Claude" when first reading.  lol

I love cuckoo filters, but two concerns:  

It wants you to put one into on-chain storage?  That's odd.  It probably doesn't understand what a blockchain is.  You'd probably keep the filter locally and make every registration checker sync then chain.  If you do not want them syncing then that's another issue.

Also cuckoo filters would usually be more vulnerable to bad hashing than typically hash maps, precisely because they waste less space.  You need to be careful when seding them between untrusting parties.
